### PR TITLE
Add migration sticker and unify the loading experience for the upgrade plan step

### LINKED ISSUE #91788: Migration Flow Improvements: Mark the blog as the one being migrated
### Details

When [the REST API endpoints](https://github.com/Automattic/dotcom-forge/issues/7759) are ready, we need to mark the blog as being migrated.

We want to call the corresponding API when we arrive at the Upgrade step, but before retrieving plans.

We have to make these two calls synchronous, as the result of getting plans depends on our marking.

It feels safer to do both calls on this step to avoid further side-effects in case other steps have changed.

We want to have a [feature flag](https://github.com/Automattic/wp-calypso/pull/92147/commits/ed8650d7a4f9ab497efb81d91d30d43763091c29) to disable this behavior.

### Checklist

_No response_

### Related

_No response_

### PR DESCRIPTION
**Pull request by renatho**
Related to #91788

## Proposed Changes

* Add a sticker marking the blog as a migration block when it's loading the upgrade plan step of the migration flow.
* I took the opportunity to unify the loaders of the upgrade plan step, preventing the flickering.
* It doesn't handle request errors while adding the sticker. Should we handle it somehow? Or just proceed with the normal price as workaround? If we want to handle it, I thought of adding some warning or something like this above the plans.

## Why are these changes being made?

* The migration sticker should be added to the blog before proceeding to the checkout step to mark it as a blog that is being migrated.

## Testing Instructions

* Enable the feature flag `migration-flow/introductory-offer`. You can add it to your `.env` file like this `ACTIVE_FEATURE_FLAGS="migration-flow/new-migration-instructions-step,migration-flow/introductory-offer"
`
* Start monitoring your browser requests.
* Start a migration and navigate until the step `site-migration-upgrade-plan`.
* Check that when that step loads, it requests the endpoint `.../{SITE_ID}/migration-flow`

## Recording

https://github.com/Automattic/wp-calypso/assets/876340/96db431a-73f2-4f80-a531-f1b5e73c83a8

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by Imran92** [comment]
As I'm working on the deleteing flag issue in parallel, I am using your branch as base. Here is the draft WIP https://github.com/Automattic/wp-calypso/pull/92172. Please LMK if you have any comment about my approach or in case you'd prefer if I create another hook for deleting.

Thanks!!!

**Comment by renatho** [comment]
> Please LMK if you have any comment about my approach or in case you'd prefer if I create another hook for deleting.

I think it's great to use the same hook! I was thinking something similar! Thank you, Imran!

File: client/blocks/importer/wordpress/upgrade-plan/style.scss
Comment (renatho) [inline client/blocks/importer/wordpress/upgrade-plan/style.scss:330]: It is not used anymore because we are unifying all loaders from the upgrade plan step.

File: client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx
Comment (renatho) [inline client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:36]: It's done in the parent component now to control the fetching along with the others.

File: client/blocks/importer/wordpress/upgrade-plan/with-migration-sticker.tsx
Comment (renatho) [inline client/blocks/importer/wordpress/upgrade-plan/with-migration-sticker.tsx:8]: I started doing it as a wrapper component, but I decided to go with a HOC in this case, which fits better because `UpgradePlanDetails` totally depends on it and it doesn't make sense using the `UpgradePlanDetails` without the wrapper (now as a HOC).

HOCs are not commonly used anymore, but in this case that we want to avoid the other hooks being called until the migration sticker is added, it will work better.

Commit changing it to a HOC: https://github.com/Automattic/wp-calypso/pull/92147/commits/7ce732245e3816b16bbdd5bb7923beb48174e475

File: config/test.json
Comment (renatho) [inline config/test.json:85]: For now, only the tests feature flag is enabled. So we can write the tests for the new feature.

File: client/blocks/importer/wordpress/import-everything/pre-migration/test/index.test.tsx
Comment (renatho) [inline client/blocks/importer/wordpress/import-everything/pre-migration/test/index.test.tsx:121]: Since I had this change here, I just changed the other mocks to also use the `as jest.Mock`, similar to other parts of the codebase, so we don't need to add the lint ignore.

File: client/blocks/importer/wordpress/import-everything/pre-migration/test/index.test.tsx
Comment (renatho) [inline client/blocks/importer/wordpress/import-everything/pre-migration/test/index.test.tsx:139]: Similar to other tests that were fixed, it now needs to wait because of the request (sticker creation) that happens before loading the component.

File: client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-loader.tsx
Comment (Imran92) [inline client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-loader.tsx]: Overall the PR is looking great and working as expected. I'm still reviewing it, but I wanted to share a little concern I have. 

Should we have a second opinion/design approval before introducing the spinner loader? Because, from what I understand from the design of different flows we have, we seem to have an strategy to actively avoid showing any loader/spinner. Rather we show what we immediately have, and update the data if new data becomes available later. Or, we show [content placeholders](https://www.codepicky.com/content-placeholders/) with fade-in/out animations, especially in the flows. For example, this comment - 
p1464106289005032-slack-C02Q9GACF . Though it's not something breaking or super serious, just wanted to share to learn about your thoughts on it

Comment (renatho) [reply client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-loader.tsx]: Thank you for the comment, Imran! Pinging @JanaMW27 to get her opinion here to see if it's not good enough and if we should try something different.

Jana, basically now we have a request to make before loading the plans for this page. So I took the opportunity to create a single loader and load everything we need for this screen, so we avoid the screen flickering while loading the dynamic content.

You can see the video attached in the PR description to see how it's working in this PR.

Comment (donnapep) [reply client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-loader.tsx]: I tend to agree with Imran's assessment. Can you use the same strategy you did for the metrics?

> so we avoid the screen flickering while loading the dynamic content.

Can you explain why the screen would flicker? It doesn't do that for the metrics.

Comment (donnapep) [reply client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-loader.tsx]: I see that on some other steps we do as in the following screenshot. For example, this is displayed after clicking the _Check my site_ button. So if we wanted to use a loader this would be the way to do it to keep things consistent, but we would need to come up with some text for this step. To keep things simple, using placeholders may be the better option:

Screenshot 2024-07-03 at 8 45 31 AM

**Comment by JanaMW27** [comment]
We have connected over slack, and we are going to use a skeleton with some animation

Screenshot 2024-07-03 at 15 40 35

Figma: https://www.figma.com/design/QnzGpMHVpKuf0ZeRQVcxiL/Migrations?node-id=7768-4939&t=0LGAfjWMGTtf4qhj-1

File: client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-loader.tsx
Comment (renatho) [reply client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-loader.tsx]: > Can you explain why the screen would flicker? It doesn't do that for the metrics.

I recorded a video here, which is better than explaining with words. =)

https://github.com/Automattic/wp-calypso/assets/876340/42c5a58f-bbf6-45ed-8742-ead1348dbcf1

So since we now need the request before loading the plans, we can join everything in a single loader, and when it's done everything goes directly to the correct place.

For the design, Jana proposed [a solution](https://github.com/Automattic/wp-calypso/pull/92147#issuecomment-2206111627) that I'll apply. 😉

File: client/blocks/importer/wordpress/upgrade-plan/skeleton/index.tsx
Comment (renatho) [inline client/blocks/importer/wordpress/upgrade-plan/skeleton/index.tsx:6]: Some wrapper elements were duplicated from the original components in order to simulate the columns.

The other way would be to add the Skeleton inside the components, but it would make it spread and demand a big refactor that wouldn't be worth it IMO.

File: client/blocks/importer/wordpress/upgrade-plan/skeleton/index.tsx
Comment (renatho) [inline client/blocks/importer/wordpress/upgrade-plan/skeleton/index.tsx:10]: Since the spacings and sizes are very dynamic to simulate a real content here, I just created the styles with default classes, and other styles are defined directly in the component.

File: config/development.json
Comment (renatho) [inline config/development.json:145]: I'm also reordering this item to have the correct order.

File: client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-loader.tsx
Comment (renatho) [reply client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-loader.tsx]: Changes done in  https://github.com/Automattic/wp-calypso/pull/92147/commits/cf2ce5bec8ee0286cb159f7913371b7c21afbc2f and https://github.com/Automattic/wp-calypso/pull/92147/commits/fe27395c5704c00ffb15f74f8f14bc68cf0432c4

It's not exactly like the design in these points:

- The left column content is not centered after loaded, so I kept the skeleton also aligned on top.
- The left column is bigger instead of the opposite like the current page is.
- I also added a skeleton for the tabs, so the containers start in the correct position when the content is loaded.
- To simplify a bit the code, the items in the right column have the same width.

I also updated the recording in the PR description.

**Review by Imran92 — APPROVED** [review]
Looks very nice! Thanks @renatho! <3

One very minor thing I've noticed is once I click on the back button, A popup comes up, but the page starts reloading in the background every time. It's probably not a big thing, if it's hard to address to investigate here, maybe we can just create a small separate issue for it

https://github.com/Automattic/wp-calypso/assets/6820724/5c1d7e0a-f95f-4c54-92fc-56e6ca47702f

**Comment by renatho** [comment]
Good catch, @Imran92! It seems, for some reason, the step is re-rendering, so it's also making another request to create the block sticker.

I'll fix it before merging this.

**Comment by renatho** [comment]
@Imran92, it was fixed with this commit: https://github.com/Automattic/wp-calypso/pull/92147/commits/b44f6f591a339ef0780f42b5d2c75280755776aa

I also removed the fade-in to avoid weird behaviors. Notice that the checkout also doesn't have the fade-in.

A very interesting thing that happens is: if we run the `useLayoutEffect` there, even with the dependencies as `[]`, it's re-running every time we open or close the modal. If seems it's happening in the `Boot` component because when we open the modal we [navigate](https://github.com/Automattic/wp-calypso/blob/b79d5972ad245a37f855bd6d6bd55cb33b844d30/client/landing/stepper/declarative-flow/site-migration-flow.ts#L411-L415), somehow triggering the `Suspense`.

See the components through the React profiler after clicking on the "Back" button:

Screenshot 2024-07-04 at 18 00 51
Screenshot 2024-07-04 at 18 01 02
Screenshot 2024-07-04 at 18 01 20


What was very interesting to me is that it's not happening with the `useEffect`, probably because it's ignoring part of the renders that are happening because the `useEffect` happens later, but this difference is unexpected for me. Maybe it's something specific related to the `Suspense`. I didn't continue digging deeper into that, but I think if we find more weird behaviors in the future we should investigate more.

**Review by merkushin — APPROVED** [review]
Looks good 👍

**Comment by Imran92** [comment]
> A very interesting thing that happens is: if we run the `useLayoutEffect` there, even with the dependencies as `[]`, it's re-running every time we open or close the modal. If seems it's happening in the `Boot` component because when we open the modal we [navigate](https://github.com/Automattic/wp-calypso/blob/b79d5972ad245a37f855bd6d6bd55cb33b844d30/client/landing/stepper/declarative-flow/site-migration-flow.ts#L411-L415), somehow triggering the `Suspense`.

That surprised me as well. After my last comment issue above, I did a quick check myself and found the triggering by navigation part. So I assumed that's why it's happening. But If that was the case, I thought it'd happen for both `useEffect` and `useLayoutEffect`. I'll look into it in my free time to understand what's happening here. But overall, I feel better that we've switched to `useEffect` as it's not blocking the UI, so hopefully lesser potential for performance issue.

**Review by Imran92 — APPROVED** [review]

**Pull request merged**
